File: DataStructures/6_Graph/csharp/graph-implementation-app/graph-implementation.cs
using System;
using System.Collections.Generic;
using System.Text;


public class Graph<T> where T : notnull
{
  //hash table to store adjacency list
  private Dictionary<T, List<T>> _adjacencyList;
  private int _numberOfNodes;

  public Graph()
  {
    this._adjacencyList = new Dictionary<T, List<T>>();
    _numberOfNodes = 0;
  }

  public void AddVertex(T nodeKey)
  {
    if (_adjacencyList.ContainsKey(nodeKey))
    {
      Console.WriteLine("Key is already in the graph");
      // Throw an exception, return, or handle it in another way based on your requirements.
      return;
    }

    _adjacencyList.Add(nodeKey, new List<T>());
    _numberOfNodes++;

  }
  public void AddEdge(T node1Key, T node2Key)
  {
    if (!_adjacencyList.ContainsKey(node1Key)
    || !_adjacencyList.ContainsKey(node2Key))
    {
      Console.WriteLine("Error: One or more nodes do not exist in the graph.");
      return;
    }

    if (_adjacencyList[node1Key].Contains(node2Key)
    && _adjacencyList[node2Key].Contains(node1Key))
    {
      Console.WriteLine($"Error: Edge between {node1Key} and {node2Key} already exists.");
      return;
    }

    //undirected graph
    _adjacencyList[node1Key].Add(node2Key);
    _adjacencyList[node2Key].Add(node1Key);
  }

  public void ShowConnections()
  {
    foreach (var node in _adjacencyList)
    {
      var nodeConnections = node.Value;
      StringBuilder connections = new StringBuilder();
      connections.Append($"{node.Key} ---> ");
      foreach (var edge in nodeConnections)
      {
        connections.Append($"{edge}  ");
      }
      Console.WriteLine(connections);
    }
  }
}

public class Program
{
  public static void Main(string[] args)
  {
    var myGraph = new Graph<string>();
    myGraph.AddVertex("0");
    myGraph.AddVertex("1");
    myGraph.AddVertex("2");
    myGraph.AddVertex("3");
    myGraph.AddVertex("4");
    myGraph.AddVertex("5");
    myGraph.AddVertex("6");
    myGraph.AddVertex("6");
    myGraph.AddEdge("3", "1");
    myGraph.AddEdge("3", "4");
    myGraph.AddEdge("4", "2");
    myGraph.AddEdge("4", "5");
    myGraph.AddEdge("5", "6");
    myGraph.AddEdge("1", "2");
    myGraph.AddEdge("1", "0");
    myGraph.AddEdge("2", "0");
    myGraph.AddEdge("2", "0");

    myGraph.ShowConnections();
  }
}

File: DataStructures/6_Graph/javascript/graph-implementation.js
class Graph {
  constructor() {
    this.numberOfNodes = 0;
    this.adjacentNodesList = {}; // Object used as a hash table to store the adjacency list.
    // We could use arrays, but removing objects not in order is less efficient and less convenient.
  }

  // addVertex(node) {
  //   this.adjacentNodes[node] = [];
  //   this.numberOfNodes++;
  // }

  addVertex(node) {
    if (!this.adjacentNodesList.hasOwnProperty(node)) {
      this.adjacentNodesList[node] = [];
      this.numberOfNodes++;
    } else {
      console.log(`Node with key ${node} already exists.`);
      // You can throw an error, return a boolean indicating success, or handle it in some other way.
    }
  }

  // addEdge(node1, node2) {
  //   //UNDIRECTED GRAPH - so connections should be added both ways
  //   this.adjacentNodesList[node1].push(node2);
  //   this.adjacentNodesList[node2].push(node1);
  // }

  addEdge(node1, node2) {
    if (
      !this.adjacentNodesList.hasOwnProperty(node1) ||
      !this.adjacentNodesList.hasOwnProperty(node2)
    ) {
      console.log("Error: One or more nodes do not exist in the graph.");
      return;
    }

    if (
      this.adjacentNodesList[node1].includes(node2) ||
      this.adjacentNodesList[node2].includes(node1)
    ) {
      console.log(`Error: Edge between ${node1} and ${node2} already exists.`);
      return;
    }

    // UNDIRECTED GRAPH - so connections should be added both ways
    this.adjacentNodesList[node1].push(node2);
    this.adjacentNodesList[node2].push(node1);
  }

  showConnections() {
    const allNodes = Object.keys(this.adjacentNodesList);

    for (let node of allNodes) {
      let nodeConnections = this.adjacentNodesList[node];
      let connections = "";
      for (let vertex of nodeConnections) {
        connections += vertex + " ";
      }
      console.log(node + "--->" + connections);
    }
  }
}

const myGraph = new Graph();
myGraph.addVertex("0");
myGraph.addVertex("1");
myGraph.addVertex("2");
myGraph.addVertex("3");
myGraph.addVertex("4");
myGraph.addVertex("5");
myGraph.addVertex("6");
myGraph.addVertex("6");
myGraph.addEdge("3", "1");
myGraph.addEdge("3", "4");
myGraph.addEdge("4", "2");
myGraph.addEdge("4", "5");
myGraph.addEdge("5", "6");
myGraph.addEdge("1", "2");
myGraph.addEdge("1", "0");
myGraph.addEdge("2", "0");
myGraph.addEdge("2", "0");

myGraph.showConnections();
